### PR TITLE
For review

### PR DESCRIPTION
**Pull request by jitendra-1217**

### CODE REVIEW

File: src/FixedBasic/Redis.php [hunk 1]
```diff
@@ -0,0 +1,66 @@
+<?php
+
+namespace Jitendra\PhpValve\FixedBasic;
+
+final class Redis extends Base
+{
+    const KEY_PREFIX = 'JPVFBR:';
```
Comment (patlola) [inline src/FixedBasic/Redis.php]: configurable ?

File: src/FixedBasic/Redis.php [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    protected $redis;
+
+    public function __construct(int $window, int $limit, \Predis\Client $redis = null)
+    {
+        parent::__construct($window, $limit);
+
+        $this->redis = $redis ?: new \Predis\Client;
+    }
+
+    public function attempt(string $resource, int $worth = 1): array
```
Comment (saurav1991) [inline src/FixedBasic/Redis.php]: why call this variable `worth`? any reference?

Comment (jitendra-1217) [reply src/FixedBasic/Redis.php]: @saurav1991 : Saw it somewhere on aws docs. For e.g. some resource are expensive than others so accessing them will cost you more.

File: src/FixedBasic/Redis.php (hunk 1 above)
Comment (jitendra-1217) [reply src/FixedBasic/Redis.php]: @patlola : Made changes. 👍

File: src/FixedBasic/Base.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
+<?php
+
+namespace Jitendra\PhpValve\FixedBasic;
```
Comment (saurav1991) [inline src/FixedBasic/Base.php]: rename `FixedBasic` to `FixedWindow` ?

File: src/Contracts/Limiter.php [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+     * - X-RateLimit-Limit
+     * - X-RateLimit-Remaining
+     * - X-RateLimit-Reset (in milliseconds)
+     * - X-RateLimit-RetryAfter
+     *
+     * @param  string      $resource
+     * @param  int|integer $worth
+     * @return array
+     */
+    public function attempt(string $resource, int $worth = 1): array;
```
Comment (saurav1991) [inline src/Contracts/Limiter.php]: why keep default value 1 in the interface?

File: src/FixedBasic/redis.lua [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+local ttl = tonumber(ARGV[1])
+local worth = tonumber(ARGV[2])
+
+-- By default remaining window is the ttl else we get it by calling ttl command
+local remaining = ttl
+
+-- Increments the key and if it was first time, issues setex command
+local hits = tonumber(redis.call('incrby', resourceKey, worth))
+if hits == worth then
+    redis.call('setex', resourceKey, ttl, worth)
```
Comment (saurav1991) [inline src/FixedBasic/redis.lua]: just call `expire` instead of `setex`,  value is already set in previous step?

File: src/LeakyBucket/Base.php [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    protected $leakRateDuration;
+
+    public function __construct(int $maxBucketSize, int $leakRateValue, int $leakRateDuration)
+    {
+        $this->maxBucketSize    = $maxBucketSize;
+        $this->leakRateValue    = $leakRateValue;
+        $this->leakRateDuration = $leakRateDuration;
+    }
+
+    public function maxBucketSize(): int
```
Comment (saurav1991) [inline src/LeakyBucket/Base.php]: prefix getters with `get`

File: src/LeakyBucket/Base.php [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        return $this->leakRateValue;
+    }
+
+    public function leakRateDuration(): int
+    {
+        return $this->leakRateDuration;
+    }
+
+    public function leakFullTime(): int
```
Comment (saurav1991) [inline src/LeakyBucket/Base.php]: add a comment describing the purpose of this function

File: src/LeakyBucket/Redis.php [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+        parent::__construct($maxBucketSize, $leakRateValue, $leakRateDuration);
+
+        $this->redis = $redis ?: new \Predis\Client;
+    }
+
+    public function attempt(string $resource, int $worth = 1): array
+    {
+        $args = [
+            file_get_contents(__DIR__ . '/redis.lua'),
+            2,
```
Comment (saurav1991) [inline src/LeakyBucket/Redis.php]: what is 2 here?

File: src/LeakyBucket/Base.php [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        return ceil($this->maxBucketSize * $this->leakRateDuration / $this->leakRateValue);
```
Comment (saurav1991) [inline src/LeakyBucket/Base.php]: add parantheses in this function

File: src/LeakyBucket/redis.lua [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+local leakRateValue = tonumber(ARGV[2])
+local leakRateDuration = tonumber(ARGV[3])
+local leakFullTime = tonumber(ARGV[4])
+local now = tonumber(ARGV[5])
+local worth = tonumber(ARGV[6])
+
+local ttlSecs = math.ceil(leakFullTime/1000)
+local retryAfter = -1
+
+-- Reads values of 2 keys - last updated and current bucket size against key
```
Comment (saurav1991) [inline src/LeakyBucket/redis.lua]: move logic to get last upfated, bucket size and leak to a separate method?

File: src/LeakyBucket/redis.lua [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+if resourceBucketSize == nil then resourceBucketSize = 0 end
+
+-- Calculates units that should have been leaked between time the key got last updated and now
+local leak = math.floor((now - resourceLastUpdated) * leakRateValue / leakRateDuration)
+resourceBucketSize = math.max(0, resourceBucketSize - leak)
+resourceLastUpdated = now
+
+-- Determine now if attempt for worth should be allowed or not. If allowed fill
+-- in worth unit in the bucket and change related vars accordingly
+local allowAttempt = resourceBucketSize + worth <= maxBucketSize
```
Comment (saurav1991) [inline src/LeakyBucket/redis.lua]:
```lua
local allowed = 0
local currentBucketSize = resourceBucketSize + worth
if currentBucketSize <= maxBucketSize then
    resourceBucketSize = currentBucketSize
    allowed = 1
else
    allowed = 0
    retryAfter = now + leakRateDuration
end
```
possible refactor?

File: src/FixedBasic/Base.php (hunk 3 above)
Comment (jitendra-1217) [reply src/FixedBasic/Base.php]: There can be more than one algorithm which has "fixed window" trait.

File: src/Contracts/Limiter.php (hunk 4 above)
Comment (jitendra-1217) [reply src/Contracts/Limiter.php]: You mean interface need not define defaults?

File: src/LeakyBucket/Redis.php (hunk 8 above)
Comment (jitendra-1217) [reply src/LeakyBucket/Redis.php]: Ref https://redis.io/commands/eval, 2 is the number of redis keys which follows in the list of args. In eval you have to specify this.

File: src/LeakyBucket/redis.lua (hunk 10 above)
Comment (jitendra-1217) [reply src/LeakyBucket/redis.lua]: Why separate method?

This script is not suppose to increase over time and is well readable(not?) in current form. 

Adding methods will unnecessarily increase call stacks (tiny but unnecessary).

File: src/FixedBasic/Redis.php [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+            1,
+            $this->keyPrefix . $resource,
+            ceil($this->window/1000),
+            $worth,
+        ];
+
+        list($hits, $ttlSecsRemaining) = $this->redis->eval(...$args);
+
+        // Following logic could easily be offloaded to LUA script below
+        $allowed    = intval($hits <= $this->limit);
```
Comment (patlola) [inline src/FixedBasic/Redis.php]: can you use boolval?

File: src/Traits/HasAttributes.php [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+ */
+trait HasAttributes
+{
+    /**
+     * Redis key prefix. Useful if multiple limiters are being
+     * used on same identifier.
+     *
+     * @var string
+     */
+    protected $keyPrefix = 'pv:';
```
Comment (saurav1991) [inline src/Traits/HasAttributes.php]: use only `prefix`

File: src/FixedBasic/Redis.php (hunk 12 above)
Comment (jitendra-1217) [reply src/FixedBasic/Redis.php]: Redis returns integer for most of the operations. Lua doesn't have typecasting and need to put expression if..else.

Don't see any harm in returning 1/0 and want to keep lua minimum.

File: src/Base/Response.php [hunk 14]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace Jitendra\PhpValve\Base;
+
+use Jitendra\PhpValve\Contracts;
+
+class Response implements Contracts\Response
+{
+    public $allowed;
```
Comment (amitmahbubani) [inline src/Base/Response.php]: please PHPDoc

File: src/FixedBasic/Redis.php [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    protected $redis;
+
+    public function __construct(int $window, int $limit, \Predis\Client $redis = null)
+    {
+        parent::__construct($window, $limit);
+
+        $this->redis = $redis ?: new \Predis\Client;
+    }
+
+    public function attempt(string $resource, int $worth = 1): Response
```
Comment (amitmahbubani) [inline src/FixedBasic/Redis.php]: s/worth/cost

File: src/FixedBasic/redis.lua [hunk 16]
```diff
@@ -0,0 +1,20 @@
+-- Gets keys and arguments from command
+local key   = KEYS[1]
+local ttl   = tonumber(ARGV[1])
+local worth = tonumber(ARGV[2])
```
Comment (amitmahbubani) [inline src/FixedBasic/redis.lua]: cost :|

File: src/FixedBasic/redis.lua [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+
+-- Increments the key and if it was first time, issues setex command
```
Comment (amitmahbubani) [inline src/FixedBasic/redis.lua]: Fix setex comment

File: src/FixedBasic/Redis.php [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+            1,
+            $this->prefix . $resource,
+            $this->window,
+            $worth,
+        ];
+
+        list($hits, $ttl) = $this->redis->eval(...$args);
+
+        // Following logic could easily be offloaded to LUA script below
+        $allowed    = intval($hits <= $this->limit);
```
Comment (amitmahbubani) [inline src/FixedBasic/Redis.php]: Why not bool?

File: tests/LeakyBucket/TestCase.php [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+            $expected = new Response(1, $limiterMaxBucketSize, $limiterMaxBucketSize - $i,  time() + $limiterLeakFullTime, -1);
+            $this->attemptAndAssert($limiter, $resource, 1, $expected);
+        }
+
+        // Subsequent attempts must fail withing current leak duration
+        $expected = new Response(0, $limiterMaxBucketSize,  0,  time() + $limiterLeakFullTime,  time() + $limiterLeakRateDuration);
+        $this->attemptAndAssert($limiter, $resource, 1, $expected);
+
+        // Sleeping for while will leak and allow for new requests
+        sleep($limiterLeakRateDuration);
```
Comment (amitmahbubani) [inline tests/LeakyBucket/TestCase.php]: Not good, needs mock

File: src/Base/Response.php [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+    public $retryAfter;
+
+    public function __construct(
+        int $allowed,
+        int $limit,
+        int $remaining,
+        int $resetAt,
+        int $retryAfter)
+    {
+        $this->allowed    = $allowed;
```
Comment (amitmahbubani) [inline src/Base/Response.php]: Make this boolean.

**Comment by jitendra-1217** [comment]
Closing this.

Earlier comments has been resolved. And this pr is approved by @amitmahbubani , @saurav1991 .

Last few comments are fixed in separate pr.

**Pull request closed without merging**